blog/2024-06-02-kqir-query-engine/index.md

```diff
@@ -0,0 +1,222 @@
+
+And KQIR is a multi-level IR that can represent query structures at various levels during optimization.
+The initial transformation from the syntax tree results in Syntactical IR, a high-level representation of certain syntactic expressions.
+As it undergoes processing by an IR optimizer, KQIR evolves into Planning IR, a low-level representation used to express query execution plans within the query engine.
+
+Additionally, we will conduct semantic checks on the IR before optimization to ensure that the query is semantically correct.
+This includes verifying that it does not include any undefined schemas or fields and uses the appropriate field types.
+
+### IR Optimizer
+
+The KQIR optimizer consists of multiple passes, [a concept borrowed from LLVM](https://llvm.org/docs/Passes.html).
+Each pass takes IR as input, conducts analysis and modifications, and generates a new IR.
+
+Currently, the optimizer's passes are categorized into three main groups:
+- expression passes for optimizing logical expressions like `AND`, `OR`, `NOT` operators;
+- numeric passes for optimizing numerical comparisons with an interval analysis (i.e. analyze the mathematical properties of numerical comparisons in terms of intervals) to enhance query optimization by eliminating unnecessary comparisons or improving comparison expressions;
+- planning passes for converting syntactical IR to planning IR and enhancing query plans through a cost model that selects optimal indexes and removes unnecessary sortings.
+
+Pass execution order is controlled by the pass manager.
+A pass may run multiple times at different stages to simplify individual passes by combining them.
+
+### Plan Executor
+
+The KQIR plan executor is built on the Volcano model.
+
+Once the IR optimizer finishes all optimizations, the resulting IR becomes a planning IR. This will then be passed to the plan executor to create execution logic based on certain context corresponding to the plan operator.
+
+Subsequently, Kvrocks retrieves query results through iterative execution.
+
+### On-disk indexing
+
+Unlike Redis, which stores index data in memory, Kvrocks requires the construction of indexes on the disk.
+This means that for any field type (e.g. tag, numeric), we need an encoding to reduce such index to rocksdb key-values.
+
+Furthermore, we incrementally create indexes before and after JSON or HASH commands getting executed to guarantee that query results are in real-time.
+
+## Current status
+
+The KQIR functionality is currently available on the `unstable` branch, supporting commands like `FT.CREATE`, `FT.SEARCH`, and `FT.SEARCHSQL` (an extension for running SQL queries) to encourage user to test.
+
+However, as KQIR is still in early development, compatibility cannot be guaranteed and many features remain incomplete.
+Thus the upcoming release (version 2.9.0) will not include any KQIR component.
```

Comment:
Because KQIR's transaction support is very poor, users may encounter some very strange problems when querying (such as consistency issues), and there are many error conditions that I have not considered at present, which may make users have a bad impression of KQIR.
I hope to wait a little longer before releasing it.